File: count_loc.js

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

function countLines() {
  let files;
  try {
    files = execSync('git ls-files', { encoding: 'utf-8' }).split('\n').filter(Boolean);
  } catch (e) {
    console.error('Error running git:', e);
    return;
  }

  const extensions = {
    '.py': 'Python',
    '.ts': 'TypeScript',
    '.tsx': 'TypeScript (React)',
    '.js': 'JavaScript',
    '.jsx': 'JavaScript (React)',
    '.css': 'CSS',
    '.less': 'Less',
    '.html': 'HTML',
    '.md': 'Markdown',
    '.yml': 'YAML',
    '.yaml': 'YAML',
    '.sql': 'SQL',
    '.json': 'JSON'
  };

  const stats = {};
  Object.values(extensions).forEach(lang => {
    stats[lang] = { files: 0, lines: 0 };
  });
  stats['Other'] = { files: 0, lines: 0 };

  let totalLines = 0;
  let totalFiles = 0;

  files.forEach(f => {
    if (!fs.existsSync(f)) return;
    try {
      if (fs.lstatSync(f).isDirectory()) return;
    } catch (e) { return; }
    
    if (f.includes('node_modules')) return;

    const ext = path.extname(f);
    const lang = extensions[ext] || 'Other';

    try {
      // Using readFileSync with encoding might fail for binary files, so we handle it
      const content = fs.readFileSync(f, 'utf-8');
      const lines = content.split('\n').length;
      stats[lang].lines += lines;
      stats[lang].files += 1;
      totalLines += lines;
      totalFiles += 1;
    } catch (e) {
      // Probably a binary file or too large, skip
    }
  });

  console.log(`${'Language'.padEnd(25)} | ${'Files'.padEnd(10)} | ${'Lines'.padEnd(10)}`);
  console.log('-'.repeat(50));
  Object.entries(stats)
    .sort((a, b) => b[1].lines - a[1].lines)
    .forEach(([lang, data]) => {
      if (data.lines > 0) {
        console.log(`${lang.padEnd(25)} | ${String(data.files).padEnd(10)} | ${String(data.lines).padEnd(10)}`);
      }
    });

  console.log('-'.repeat(50));
  console.log(`${'Total'.padEnd(25)} | ${String(totalFiles).padEnd(10)} | ${String(totalLines).padEnd(10)}`);
}

countLines();

File: count_loc.py

import os
import subprocess

def count_lines():
    try:
        # Get all tracked files using git
        files = subprocess.check_output(['git', 'ls-files']).decode('utf-8').splitlines()
    except Exception as e:
        print(f"Error running git: {e}")
        return

    extensions = {
        '.py': 'Python',
        '.ts': 'TypeScript',
        '.tsx': 'TypeScript (React)',
        '.js': 'JavaScript',
        '.jsx': 'JavaScript (React)',
        '.css': 'CSS',
        '.less': 'Less',
        '.html': 'HTML',
        '.md': 'Markdown',
        '.yml': 'YAML',
        '.yaml': 'YAML',
        '.sql': 'SQL'
    }

    stats = {lang: {'files': 0, 'lines': 0} for lang in set(extensions.values())}
    stats['Other'] = {'files': 0, 'lines': 0}

    total_lines = 0
    total_files = 0

    for f in files:
        if not os.path.isfile(f):
            continue
        
        # Skip node_modules or other giant vendor dirs if they somehow got tracked
        if 'node_modules' in f:
            continue

        ext = os.path.splitext(f)[1]
        lang = extensions.get(ext, 'Other')
        
        try:
            with open(f, 'r', encoding='utf-8', errors='ignore') as file:
                lines = sum(1 for line in file)
                stats[lang]['lines'] += lines
                stats[lang]['files'] += 1
                total_lines += lines
                total_files += 1
        except Exception:
            continue

    print(f"{'Language':<25} | {'Files':<10} | {'Lines':<10}")
    print("-" * 50)
    for lang, data in sorted(stats.items(), key=lambda x: x[1]['lines'], reverse=True):
        if data['lines'] > 0:
            print(f"{lang:<25} | {data['files']:<10} | {data['lines']:<10}")
    
    print("-" * 50)
    print(f"{'Total':<25} | {total_files:<10} | {total_lines:<10}")

if __name__ == "__main__":
    count_lines()
